eng/pipelines/templates/steps/analyze.yml

```diff
@@ -0,0 +1,121 @@
+parameters:
+  - name: Artifacts
+    type: object
+    default: []
+  - name: ServiceDirectory
+    type: string
+    default: not-specified
+  - name: SDKType
+    type: string
+    default: all
+  - name: BuildSnippets
+    type: boolean
+    default: true
+
+steps:
+  - ${{ if not(contains(variables['Build.DefinitionName'], '-pr - ')) }}:
+    - template: /eng/common/pipelines/templates/steps/sparse-checkout.yml
+      parameters:
+          Paths:
+            - "/*"
+            - "!SessionRecords"
+
+  - template: /eng/pipelines/templates/steps/install-dotnet.yml
+
+  - template: /eng/common/pipelines/templates/steps/cache-ps-modules.yml
+
+  - template: /eng/common/pipelines/templates/steps/save-package-properties.yml
+    parameters:
+      ServiceDirectory: ${{ parameters.ServiceDirectory }}
+
+  - template: /eng/pipelines/templates/steps/set-artifact-packages.yml
+    parameters:
+      PackageInfo: '$(Build.ArtifactStagingDirectory)/PackageInfo'
+      Artifacts: ${{ parameters.Artifacts }}
+
+  - template: /eng/common/pipelines/templates/steps/verify-samples.yml
+    parameters:
+      ServiceDirectories: "$(ChangedServices)"
+
+  - template: /eng/common/pipelines/templates/steps/verify-readmes.yml
+    parameters:
+      PackagePropertiesFolder: '$(Build.ArtifactStagingDirectory)/PackageInfo'
+
+  - task: NodeTool@0
+    inputs:
+      versionSpec: 18.x
+    displayName: 'Use Node 18.x'
+
+  - task: DotNetCoreCLI@2
+    displayName: 'Install Snippet Generator Tool'
+    inputs:
+      command: custom
+      custom: 'tool'
+      arguments: 'restore'
+      workingDirectory: '$(Agent.BuildDirectory)'
+
+  - pwsh: |
+      $failed = $false
+      if ("$(ProjectNames)") {
+        $serviceDirectories = "$(ChangedServices)" -split ","
+        $ScanPaths = @()
+        foreach ($ServiceDirectory in $serviceDirectories) {
+          ./eng/scripts/CodeChecks.ps1 `
+            -ServiceDirectory $ServiceDirectory `
+            -SDKType ${{ parameters.SDKType }} `
+            -SpellCheckPublicApiSurface `
+
+          if ($LASTEXITCODE -ne 0) {
+            $failed = $true
+          }
+        }
+
+        if ($failed) {
+          exit 1
+        }
+      }
+      else {
+        Write-Host "No services were directly changed. Skipping code checks."
+      }
+    displayName: Verify Generated Code
+    env:
+      EnableSourceLink: false
+
+  - template: /eng/pipelines/templates/steps/dotnet-diagnostics.yml
+    parameters:
+      LogFilePath: $(Build.ArtifactStagingDirectory)/rebuild.binlog
+
+  # todo, refactor to using the projectlistoverridefile
```

Comment:
I assume this todo is left to be completed?

Reply:
Yes absolutely. This will get properly handled in the build parameters pr I've got open. I'll remove the todo from this PR as it's just confusing as hell.